internal/db/gc.go

```diff
@@ -0,0 +1,98 @@
+package db
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgtype"
+)
+
+func GcProjectObjects(ctx context.Context, tx pgx.Tx, project int64, keep int64, fromVersion int64) ([]Hash, error) {
+	hashes := []Hash{}
+
+	rows, err := tx.Query(ctx, `
+		WITH latest AS (
+			SELECT latest_version AS version
+			FROM dl.projects
+			WHERE id = $1
+		)
+		DELETE FROM dl.objects
+		WHERE project = $1
+		  AND start_version > $2
+		  AND stop_version IS NOT NULL
+		  AND stop_version <= ((SELECT version FROM latest) - $3)
+		RETURNING path, (hash).h1, (hash).h2
+	`, project, fromVersion, keep)
+	if err != nil {
+		return hashes, fmt.Errorf("GcProjectObjects query, project %v, keep %v, from %v: %w", project, keep, fromVersion, err)
+	}
+
+	for rows.Next() {
+		var path string
+		var hash Hash
+		err = rows.Scan(&path, &hash.H1, &hash.H2)
+		if err != nil {
+			return hashes, fmt.Errorf("GcProjectObjects scan %v: %w", project, err)
+		}
+
+		hashes = append(hashes, hash)
+	}
+
+	return hashes, nil
+}
+
+func GcContentHashes(ctx context.Context, tx pgx.Tx, hashes []Hash) (int64, error) {
+	if len(hashes) == 0 {
+		return 0, nil
+	}
+
+	// FIXME: Do this on server boot
+	hashType, err := tx.Conn().LoadType(ctx, "hash")
+	if err != nil {
+		return 0, fmt.Errorf("load hash type: %w", err)
+	}
+	tx.Conn().TypeMap().RegisterType(hashType)
+
+	var params strings.Builder
+	for idx := 1; idx <= len(hashes); idx++ {
+		params.WriteString(fmt.Sprintf("$%d::hash", idx))
+		if idx != len(hashes) {
+			params.WriteRune(',')
+		}
+	}
+
+	query := fmt.Sprintf(`
+		WITH missing AS (
+			SELECT c.hash
+			FROM dl.contents c
+			LEFT JOIN dl.objects o
+				ON c.hash = o.hash
+			WHERE c.hash IN (%s)
```

Comment:
We talked about this IRL but just to document our thoughts.

It would be nice if we used an array parameter here instead of a list of dynamic ones. Otherwise, PGAnalyze will think all of these queries are unique and we won't get nice metrics.